dali/operators/random/rng_base.h

```diff
@@ -0,0 +1,130 @@
+#include "dali/core/convert.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/util/batch_rng.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+
+DALI_SCHEMA(RNGAttr)
+    .DocStr(R"code()code")
+    .AddOptionalArg<std::vector<int>>("shape",
+      R"code(Shape of the data.)code", nullptr, true)
+    .AddOptionalArg<DALIDataType>("dtype",
+      R"code(Data type.)code", nullptr);
+
+template<typename Backend, typename Impl>
+class RNGBase : public Operator<Backend> {
+ public:
+  ~RNGBase() override = default;
+
+ protected:
+  explicit RNGBase(const OpSpec &spec)
+      : Operator<Backend>(spec),
+        rng_(spec.GetArgument<int64_t>("seed"), spec.GetArgument<DALIDataType>("batch_size")) {
+  }
+
+  Impl &This() noexcept { return static_cast<Impl&>(*this); }
+  const Impl &This() const noexcept { return static_cast<const Impl&>(*this); }
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc,
+                 const workspace_t<Backend> &ws) override {
+    if (!spec_.TryGetArgument(dtype_, "dtype"))
+      dtype_ = This().DefaultDataType();
+
+    bool has_shape = spec_.ArgumentDefined("shape");
+    bool has_shape_like = spec_.NumRegularInput() == 1;
+    DALI_ENFORCE(!(has_shape && has_shape_like),
+      "Providing argument \"shape\" is incompatible with providing a shape-like input");
+
+    if (has_shape_like) {
+      shape_ = ws.template InputRef<Backend>(0).shape();
+    } else if (has_shape) {
+      GetShapeArgument(shape_, spec_, "shape", ws);
+    } else {
+      shape_ = uniform_list_shape(spec_.template GetArgument<DALIDataType>("batch_size"), {1});
+    }
+    batch_size_ = shape_.size();
+
+    This().AcquireArgs(spec_, ws, batch_size_);
+
+    output_desc.resize(1);
+    output_desc[0].shape = shape_;
+    output_desc[0].type = TypeTable::GetTypeInfo(dtype_);
+    return true;
+  }
+
+  template <typename T>
+  void RunImplTyped(workspace_t<CPUBackend> &ws) {
+    auto &output = ws.OutputRef<CPUBackend>(0);
+    auto out_view = view<T>(output);
+    const auto &out_shape = out_view.shape;
+    auto &tp = ws.GetThreadPool();
+    constexpr int64_t kThreshold = 0; 1 << 18;
+    constexpr int64_t kChunkSize = 1 << 16;
+    constexpr size_t kNumChunkSeeds = 4;
+
+    for (int sample_id = 0; sample_id < batch_size_; ++sample_id) {
+      auto sample_sz = out_shape.tensor_size(sample_id);
+      span<T> out_span{out_view[sample_id].data, sample_sz};
+      if (sample_sz < kThreshold) {
+        tp.AddWork(
+          [=](int thread_id) {
+            This().template Generate<T>(out_span, sample_id, rng_[sample_id]);
+          }, sample_sz);
+      } else {
+        int chunks = out_span.size() / kChunkSize;
+        std::array<uint32_t, kNumChunkSeeds> seed;
+        for (int c = 0; c < chunks; c++) {
+          auto start = out_span.begin() + c * kChunkSize;
+          auto sz = std::min(kChunkSize, out_span.end() - start);
```

Comment:
This is a bug.
Consider:
```
kChunkSize = 65536;
out_span.size() = 100000;
chunks = out_span.size() / kChunkSize = 1;
```
There's only one chunk now and the last chunk will be lost.
Assuming that chunks is calcualted as: `chunks = (out_span.size() + kChunkSize - 1) / kChunkSize`

Now consider `out_span.size() = 70000`
With start calculated as out_span.begin() + c * kChunkSize you have one chunk of 65536 and on chunk of 4464 bytes - quite an imbalance.

I suggest to equally divide the length:
```
start = out_span.size() * c / chunks;
end = out_span.size() * (c + 1) / chunks;
```
The chunks will differ by 1 elements at most.